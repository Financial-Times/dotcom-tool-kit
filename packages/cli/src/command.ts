import type { Plugin } from './plugin'

export interface CommandClass {
   id?: string
   plugin?: Plugin
   description: string
   hidden?: boolean
   new(argv: string[]): Command
}

export interface Command {
   config?: Object
   init?(): Promise<void>
   run(): Promise<void>
}
