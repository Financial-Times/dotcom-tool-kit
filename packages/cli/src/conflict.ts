import type { Plugin } from './plugin'

export interface Conflict<T> {
  plugin: Plugin
  conflicting: T[]
}

export function isConflict<T>(thing: unknown): thing is Conflict<T> {
  return Boolean((thing as Conflict<T>).conflicting)
}

export function findConflicts<T, U>(items: (U | Conflict<T>)[]): Conflict<T>[] {
  const conflicts: Conflict<T>[] = []

  for (const item of items) {
    if (isConflict<T>(item)) {
      conflicts.push(item)
    }
  }

  return conflicts
}


export function withoutConflicts<T, U>(items: (U | Conflict<T>)[]): U[] {
   const nonConflicts:U[] = []

   for(const item of items) {
      if(!isConflict<T>(item)) {
         nonConflicts.push(item)
      }
   }

   return nonConflicts
}
