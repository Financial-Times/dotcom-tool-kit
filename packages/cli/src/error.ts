export class ToolKitError extends Error {
  details?: string
}
