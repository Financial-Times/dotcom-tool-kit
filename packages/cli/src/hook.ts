import type { Plugin } from './plugin'

export interface HookTask {
  id: string
  plugin: Plugin
  tasks: string[]
}

export interface HookClass {
  id?: string
  plugin?: Plugin
  description?: string
  new (): Hook
}

export interface Hook {
  check(): Promise<boolean>
  install(): Promise<void>
}
