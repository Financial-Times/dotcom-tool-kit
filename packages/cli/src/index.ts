import { cli } from 'cli-ux'
import { error, exit } from '@oclif/errors'
import { cosmiconfig } from 'cosmiconfig'
import importFrom from 'import-from'
import resolveFrom from 'resolve-from'

const explorer = cosmiconfig('toolkit')

type Config = {
   plugins: { [id: string]: Plugin },
   commands: { [id: string]: CommandClass },
}

const config: Config = {
   plugins: {},
   commands: {}
}

interface CommandClass {
   new(argv: string[]): Command
}

interface Command {
   run(): Promise<void>
}

interface Plugin {
   id: string
   root: string
   commands: {
      [id: string]: CommandClass
   }
}

async function findToolKitPlugins(root: string): Promise<string[]> {
  const result = await explorer.search(root)
  if (!result) return []

  const { plugins = [] } = result.config

  return plugins
}

async function loadPlugin(id: string, root: string): Promise<Plugin> {
   // don't load duplicate commands
   if(id in config.plugins) {
      return config.plugins[id]
   }

   // load plugin relative to the app or parent plugin
   const pluginRoot = resolveFrom(root, id)
   const plugin = importFrom(root, id) as Plugin

   config.plugins[id] = plugin
   plugin.id = id
   plugin.root = pluginRoot

   // TODO check duplicate commands
   // TODO lifecycles
   Object.assign(
      config.commands,
      plugin.commands
   )

   // load any plugins requested by this plugin
   await loadPlugins(plugin.root)

   // TODO run init hook(? or just put init code in the plugin entry point)

   return plugin
}

async function loadPlugins(root = process.cwd()) {
   const plugins = await findToolKitPlugins(root)

   return plugins.map(
      plugin => loadPlugin(plugin, root)
   )
}

function validatePlugins() {
   // TODO reimplement
   const duplicateCommands: [string, string[]][] = []

  if(duplicateCommands.length !== 0) {
    console.log(`Error: you have multiple plugins installed that have conflicting commands, which isn't allowed. remove all but one of these plugins from your app's package.json:\n`)

    cli.table(duplicateCommands, {
      plugins: { get: row => row[1].join(', ') + '   ' },
      command: { get: row => row[0] },
    })

    exit(1)
  }
}

export async function load() {
  await loadPlugins()
  validatePlugins()
  return config
}

export async function runCommand(id: string, argv: string[]) {
   // TODO checking command exists, running help
   const Command = config.commands[id]
   const command = new Command(argv)
   return command.run()
}
