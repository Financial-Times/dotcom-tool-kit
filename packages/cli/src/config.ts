import path from 'path'

import type { TaskClass } from '@dotcom-tool-kit/task'
import type { HookTask, HookClass } from './hook'
import { loadPluginConfig, Plugin } from './plugin'
import { Conflict, findConflicts, withoutConflicts, isConflict } from './conflict'
import { ToolKitConflictError, ToolKitError } from '@dotcom-tool-kit/error'
import {
  formatTaskConflicts,
  formatUndefinedHookTasks,
  formatUnusedOptions,
  formatHookTaskConflicts,
  formatHookConflicts,
  formatOptionConflicts,
  formatUninstalledHooks,
  formatMissingTasks
} from './messages'

export interface PluginOptions {
  options: Record<string, unknown>
  plugin: Plugin
  forPlugin: Plugin
}

export interface Config {
  root: string
  plugins: { [id: string]: Plugin }
  tasks: { [id: string]: TaskClass | Conflict<TaskClass> }
  hookTasks: { [id: string]: HookTask | Conflict<HookTask> }
  options: { [id: string]: PluginOptions | Conflict<PluginOptions> | undefined }
  hooks: { [id: string]: HookClass | Conflict<HookClass> }
}

export interface ValidConfig extends Config {
  tasks: { [id: string]: TaskClass }
  hookTasks: { [id: string]: HookTask }
  options: { [id: string]: PluginOptions }
  hooks: { [id: string]: HookClass }
}

const coreRoot = path.resolve(__dirname, '../')

const createConfig = (): Config => ({
  root: coreRoot,
  plugins: {},
  tasks: {},
  hookTasks: {},
  options: {},
  hooks: {}
})

async function asyncFilter<T>(items: T[], predicate: (item: T) => Promise<boolean>): Promise<T[]> {
  const results = await Promise.all(items.map(async (item) => ({ item, keep: await predicate(item) })))

  return results.filter(({ keep }) => keep).map(({ item }) => item)
}

export async function validateConfig(config: Config): Promise<ValidConfig> {
  const hookTaskConflicts = findConflicts(Object.values(config.hookTasks))
  const hookConflicts = findConflicts(Object.values(config.hooks))
  const taskConflicts = findConflicts(Object.values(config.tasks))
  const optionConflicts = findConflicts(Object.values(config.options))

  const definedHookTaskConflicts = hookTaskConflicts.filter((conflict) => {
    return conflict.conflicting[0].id in config.hooks
  })

  let shouldThrow = false
  const error = new ToolKitConflictError(
    'There are problems with your Tool Kit configuration.',
    hookTaskConflicts.map((conflict) => ({
      hook: conflict.conflicting[0].id,
      conflictingTasks: conflict.conflicting.flatMap((hook) =>
        hook.tasks.map((task) => ({ task, plugin: hook.plugin.id }))
      )
    }))
  )
  error.details = ''

  if (
    hookConflicts.length > 0 ||
    definedHookTaskConflicts.length > 0 ||
    taskConflicts.length > 0 ||
    optionConflicts.length > 0
  ) {
    shouldThrow = true

    if (hookConflicts.length) {
      error.details += formatHookConflicts(hookConflicts)
    }

    if (definedHookTaskConflicts.length) {
      error.details += formatHookTaskConflicts(definedHookTaskConflicts)
    }

    if (taskConflicts.length) {
      error.details += formatTaskConflicts(taskConflicts)
    }

    if (optionConflicts.length) {
      error.details += formatOptionConflicts(optionConflicts)
    }
  }

  const configuredHookTasks = withoutConflicts(Object.values(config.hookTasks))
  const definedHookIds = new Set(Object.keys(config.hooks))
  const undefinedHookTasks = configuredHookTasks.filter((hookTask) => {
    // we only care about undefined hooks that were configured by the app, not default config from plugins
    const fromApp = hookTask.plugin.root === process.cwd()
    const hookDefined = definedHookIds.has(hookTask.id)
    return fromApp && !hookDefined
  })

  if (undefinedHookTasks.length > 0) {
    shouldThrow = true
    error.details += formatUndefinedHookTasks(undefinedHookTasks, Array.from(definedHookIds))
  }

  const unusedOptions = Object.entries(config.options)
    .filter(
      ([, option]) =>
        option && !isConflict(option) && !option.forPlugin && option.plugin.root === process.cwd()
    )
    .map(([id]) => id)
  if (unusedOptions.length > 0) {
    shouldThrow = true
    error.details += formatUnusedOptions(unusedOptions, Object.keys(config.plugins))
  }

  const missingTasks = configuredHookTasks
    .map((hook) => ({
      hook,
      tasks: hook.tasks.filter((id) => !config.tasks[id])
    }))
    .filter(({ tasks }) => tasks.length > 0)

  if (missingTasks.length > 0) {
    shouldThrow = true
    error.details += formatMissingTasks(missingTasks, Object.keys(config.tasks))
  }

  if (shouldThrow) {
    throw error
  }

  return config as ValidConfig
}

export async function checkInstall(config: ValidConfig): Promise<void> {
  const definedHooks = withoutConflicts(Object.values(config.hooks))
  const uninstalledHooks = await asyncFilter(definedHooks, async (Hook) => {
    const hook = new Hook()
    return !(await hook.check())
  })

  if (uninstalledHooks.length > 0) {
    const error = new ToolKitError('There are problems with your Tool Kit installation.')
    error.details = formatUninstalledHooks(uninstalledHooks)
    throw error
  }
}

export function loadConfig(options?: { validate?: true }): Promise<ValidConfig>
export function loadConfig(options?: { validate?: false }): Promise<Config>

export async function loadConfig({ validate = true } = {}): Promise<ValidConfig | Config> {
  // start loading config and child plugins, starting from the consumer app directory
  const config = await loadPluginConfig(
    {
      id: 'app root',
      root: process.cwd()
    },
    createConfig()
  )

  return validate ? validateConfig(config) : config
}
