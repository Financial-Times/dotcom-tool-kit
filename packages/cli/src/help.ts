import { loadConfig } from './config'
import { styles } from './messages'

export default async function showHelp(hooks: string[]): Promise<void> {
  const config = await loadConfig({ checkInstall: false })

  if (hooks.length === 0) {
    hooks = Object.keys(config.hooks).sort()
  }

  const missingHooks = hooks.filter((hook) => !config.hooks[hook])

  console.log(`
ðŸ§° ${styles.title(`welcome to ${styles.app('Tool Kit')}!`)}

Tool Kit is modern, maintainable & modular developer tooling for FT.com projects.

${styles.URL('https://github.com/financial-times/dotcom-tool-kit')}

${styles.ruler()}
${styles.dim(
  hooks.length === 0
    ? 'available hooks'
    : `help for ${hooks.length - missingHooks.length} ${
        hooks.length - missingHooks.length > 1 ? 'hooks' : 'hook'
      }`
)}:
`)

  for (const hook of hooks) {
    const Hook = config.hooks[hook]

    if (Hook) {
      const tasks = config.hookTasks[hook]
      console.log(`${styles.heading(hook)}
${Hook.description ? Hook.description + '\n' : ''}
${
  tasks && tasks.tasks.length
    ? `runs ${tasks.tasks.length > 1 ? 'these tasks' : 'this task'}:
${tasks.tasks
  .map((task) => `- ${styles.task(task)} ${styles.dim(config.tasks[task].description)}`)
  .join('\n')}`
    : styles.dim('no tasks configured to run on this hook.')
}
${styles.ruler()}
`)
    }
  }

  if (missingHooks.length) {
    console.warn(
      styles.warning(
        `no such ${missingHooks.length > 1 ? 'hooks' : 'hook'} ${missingHooks.map(styles.hook).join(', ')}`
      )
    )
  }
}
