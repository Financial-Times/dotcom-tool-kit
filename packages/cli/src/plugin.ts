import importFrom from 'import-from'
import resolveFrom from 'resolve-from'
import mergeWith from 'lodash.mergewith'

import type { CommandClass } from './command'
import type { Lifecycle } from './lifecycle'
import { Conflict, isConflict } from './conflict'
import { config, PluginOptions } from './config'
import { loadToolKitRC } from './rc-file'

export interface Plugin {
   id: string
   root: string
   parent?: Plugin
   commands?: {
      [id: string]: CommandClass
   }
}

export async function loadPluginConfig(plugin: Plugin) {
   const { plugins = [], lifecycles = {}, options = {} } = await loadToolKitRC(plugin.root)

   // add plugin commands to our command registry, handling any conflicts
   mergeWith(
      config.commands,
      plugin.commands,

      (existingCommand: CommandClass | Conflict<CommandClass>, newCommand: CommandClass, commandId): CommandClass | Conflict<CommandClass> => {
         newCommand.plugin = plugin
         newCommand.id = commandId

         if(!existingCommand) {
            return newCommand
         }

         const conflicting = isConflict(existingCommand)
            ? existingCommand.conflicting
            : [existingCommand]

         return {
            plugin,
            conflicting: conflicting.concat(newCommand)
         }
      }
   )

   // load any plugins requested by this plugin
   await loadPlugins(plugins, plugin)

   // load plugin lifecycle assignments. do this after loading child plugins, so
   // parent lifecycles get assigned after child lifecycles and can override them
   mergeWith(
      config.lifecycles,
      lifecycles,

      // handle conflicts between lifecycles from different plugins
      (existingLifecycle: Lifecycle | Conflict<Lifecycle> | undefined, configLifecycle: string | string[], id): Lifecycle | Conflict<Lifecycle> => {
         const newLifecycle: Lifecycle = {
            id,
            plugin,
            commands: Array.isArray(configLifecycle) ? configLifecycle : [configLifecycle],
         }

         // this lifecycle might not have been set yet, in which case use the new one
         if(!existingLifecycle) {
            return newLifecycle
         }

         const existingFromSibling = existingLifecycle.plugin.parent && existingLifecycle.plugin.parent === plugin.parent

         // if the existing lifecycle was from a sibling, that's a conflict
         // return a conflict either listing this lifecycle and the siblings,
         // or merging in a previously-generated lifecycle
         if(existingFromSibling) {
            const conflicting = isConflict(existingLifecycle)
               ? existingLifecycle.conflicting
               : [existingLifecycle]

            const conflict: Conflict<Lifecycle> = {
               plugin,
               conflicting: conflicting.concat(newLifecycle)
            }

            return conflict
         }

         // if we're here, any existing lifecycle is from a child plugin,
         // so the parent always overrides it
         return newLifecycle
      }
   )

   // merge options from this plugin's config with any options we've collected already
   // TODO this is almost the exact same code as for lifecycles, refactor
   mergeWith(
      config.options,
      options,
      (existingOptions: PluginOptions | Conflict<PluginOptions>, configOptions: Object, id) => {
         const pluginOptions: PluginOptions = {
            options: configOptions,
            plugin,
            forPlugin: config.plugins[id]
         }

         // this lifecycle might not have been set yet, in which case use the new one
         if(!existingOptions) {
            return pluginOptions
         }

         const existingFromSibling = existingOptions.plugin.parent && existingOptions.plugin.parent === plugin.parent

         // if the existing options were from a sibling, that's a conflict
         // return a conflict either listing these options and the sibling's,
         // or merging in previously-generated options
         if(existingFromSibling) {
            const conflicting = isConflict(existingOptions)
               ? existingOptions.conflicting
               : [existingOptions]

            const conflict: Conflict<PluginOptions> = {
               plugin,
               conflicting: conflicting.concat(pluginOptions)
            }

            return conflict
         }

         // if we're here, any existing options are from a child plugin,
         // so merge in overrides from the parent
         return { ...existingOptions , ...pluginOptions }
      }
   )
}

export async function loadPlugin(id: string, parent?: Plugin): Promise<Plugin> {
   // don't load duplicate commands
   if (id in config.plugins) {
      return config.plugins[id]
   }

   const root = parent ? parent.root : process.cwd()

   // load plugin relative to the parent plugin
   const pluginRoot = resolveFrom(root, id)
   const basePlugin = importFrom.silent(root, id) as Plugin
   const plugin: Plugin = {
      ...basePlugin,
      id,
      root: pluginRoot,
      parent
   }

   config.plugins[id] = plugin
   await loadPluginConfig(plugin)

   return plugin
}

export function loadPlugins(plugins: string[], parent?: Plugin) {
   return Promise.all(plugins.map(
      plugin => loadPlugin(plugin, parent)
   ))
}
