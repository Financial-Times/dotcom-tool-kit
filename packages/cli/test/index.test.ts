import { describe, jest, it, expect, afterEach } from '@jest/globals'
import cloneDeep from 'lodash.clonedeep'
import * as path from 'path'
import { ToolKitError } from '../../error'
import { config, validateConfig } from '../src/config'
import { loadPluginConfig } from '../src/plugin'

// Loading all the plugins can (unfortunately) take longer than the default 2s timeout
jest.setTimeout(15000)

const baseConfig = cloneDeep(config)

describe('cli', () => {
  afterEach(() => {
    // Need to reset the config object for each test
    Object.assign(config, cloneDeep(baseConfig))
  })

  it('should load plugins correctly', async () => {
    // use a relative path for consistent snapshot across machines
    const root = path.relative(process.cwd(), path.join(__dirname, 'files/successful'))

    await loadPluginConfig({ id: 'successful test root', root })
    await validateConfig(config, { checkInstall: false })
    expect(config).toMatchSnapshot()
  })

  it('should indicate when there are conflicts', async () => {
    await loadPluginConfig({ id: 'conflicted test root', root: path.join(__dirname, 'files/conflicted') })

    expect(() => validateConfig(config, { checkInstall: false })).rejects.toBeInstanceOf(ToolKitError)
    expect(config).toHaveProperty('lifecycleAssignments.build:ci.conflicting')
    expect(config).toHaveProperty('lifecycleAssignments.build:remote.conflicting')
    expect(config).toHaveProperty('lifecycleAssignments.build:local.conflicting')
  })

  it('should succeed when conflicts are resolved', async () => {
    await loadPluginConfig({
      id: 'resolved test root',
      root: path.join(__dirname, 'files/conflict-resolution')
    })

    const validConfig = await validateConfig(config, { checkInstall: false }).catch((e) => {
      e.message += e.details
      throw e
    })
    expect(validConfig).not.toHaveProperty('lifecycleAssignments.build:local.conflicting')
    expect(validConfig.lifecycleAssignments['build:local'].commands).toEqual([
      'webpack:development',
      'babel:development'
    ])
  })
})
