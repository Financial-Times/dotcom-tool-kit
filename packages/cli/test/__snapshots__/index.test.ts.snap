// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli should load plugins correctly 1`] = `
Object {
  "commands": Object {
    "eslint": [Function],
    "help": [Function],
    "install": [Function],
    "lifecycle": [Function],
    "mocha": [Function],
    "n-test": [Function],
  },
  "findCommand": [Function],
  "lifecycleAssignments": Object {
    "test:ci": Object {
      "commands": Array [
        "mocha",
        "eslint",
      ],
      "id": "test:ci",
      "plugin": Object {
        "id": "successful test root",
        "root": "packages/cli/test/files/successful",
      },
    },
    "test:local": Object {
      "commands": Array [
        "mocha",
        "eslint",
      ],
      "id": "test:local",
      "plugin": Object {
        "id": "successful test root",
        "root": "packages/cli/test/files/successful",
      },
    },
    "test:remote": Object {
      "commands": Array [
        "n-test",
      ],
      "id": "test:remote",
      "plugin": Object {
        "commands": Object {
          "n-test": [Function],
        },
        "id": "../../../../n-test",
        "parent": Object {
          "id": "successful test root",
          "root": "packages/cli/test/files/successful",
        },
        "root": "/Users/kara.brightwell/Work/dotcom-tool-kit/packages/n-test/lib/index.js",
      },
    },
  },
  "lifecycles": Object {
    "build:ci": [Function],
    "build:local": [Function],
    "test:ci": [Function],
    "test:local": [Function],
    "test:remote": [Function],
  },
  "options": Object {
    "../../../../eslint": Object {
      "forPlugin": Object {
        "commands": Object {
          "eslint": [Function],
        },
        "id": "../../../../eslint",
        "parent": Object {
          "id": "successful test root",
          "root": "packages/cli/test/files/successful",
        },
        "root": "/Users/kara.brightwell/Work/dotcom-tool-kit/packages/eslint/lib/index.js",
      },
      "options": Object {
        "files": Array [
          "webpack.config.js",
        ],
      },
      "plugin": Object {
        "id": "successful test root",
        "root": "packages/cli/test/files/successful",
      },
    },
    "../../../../mocha": Object {
      "forPlugin": Object {
        "commands": Object {
          "mocha": [Function],
        },
        "id": "../../../../mocha",
        "parent": Object {
          "id": "successful test root",
          "root": "packages/cli/test/files/successful",
        },
        "root": "/Users/kara.brightwell/Work/dotcom-tool-kit/packages/mocha/lib/index.js",
      },
      "options": Object {
        "testDir": "./mocha_tests",
      },
      "plugin": Object {
        "id": "successful test root",
        "root": "packages/cli/test/files/successful",
      },
    },
    "../../../../n-test": Object {
      "forPlugin": Object {
        "commands": Object {
          "n-test": [Function],
        },
        "id": "../../../../n-test",
        "parent": Object {
          "id": "successful test root",
          "root": "packages/cli/test/files/successful",
        },
        "root": "/Users/kara.brightwell/Work/dotcom-tool-kit/packages/n-test/lib/index.js",
      },
      "options": Object {
        "host": "https://example.com",
      },
      "plugin": Object {
        "id": "successful test root",
        "root": "packages/cli/test/files/successful",
      },
    },
    "@dotcom-tool-kit/eslint": Object {
      "forPlugin": undefined,
      "options": Object {
        "files": "**/*.js",
      },
      "plugin": Object {
        "commands": Object {
          "eslint": [Function],
        },
        "id": "../../../../eslint",
        "parent": Object {
          "id": "successful test root",
          "root": "packages/cli/test/files/successful",
        },
        "root": "/Users/kara.brightwell/Work/dotcom-tool-kit/packages/eslint/lib/index.js",
      },
    },
    "@dotcom-tool-kit/mocha": Object {
      "forPlugin": undefined,
      "options": Object {
        "files": "test/**/*.js",
      },
      "plugin": Object {
        "commands": Object {
          "mocha": [Function],
        },
        "id": "../../../../mocha",
        "parent": Object {
          "id": "successful test root",
          "root": "packages/cli/test/files/successful",
        },
        "root": "/Users/kara.brightwell/Work/dotcom-tool-kit/packages/mocha/lib/index.js",
      },
    },
  },
  "plugins": Object {
    "../../../../circleci": Object {
      "id": "../../../../circleci",
      "lifecycles": Object {
        "build:ci": [Function],
        "test:ci": [Function],
        "test:remote": [Function],
      },
      "parent": Object {
        "id": "successful test root",
        "root": "packages/cli/test/files/successful",
      },
      "root": "/Users/kara.brightwell/Work/dotcom-tool-kit/packages/circleci/lib/index.js",
    },
    "../../../../eslint": Object {
      "commands": Object {
        "eslint": [Function],
      },
      "id": "../../../../eslint",
      "parent": Object {
        "id": "successful test root",
        "root": "packages/cli/test/files/successful",
      },
      "root": "/Users/kara.brightwell/Work/dotcom-tool-kit/packages/eslint/lib/index.js",
    },
    "../../../../mocha": Object {
      "commands": Object {
        "mocha": [Function],
      },
      "id": "../../../../mocha",
      "parent": Object {
        "id": "successful test root",
        "root": "packages/cli/test/files/successful",
      },
      "root": "/Users/kara.brightwell/Work/dotcom-tool-kit/packages/mocha/lib/index.js",
    },
    "../../../../n-test": Object {
      "commands": Object {
        "n-test": [Function],
      },
      "id": "../../../../n-test",
      "parent": Object {
        "id": "successful test root",
        "root": "packages/cli/test/files/successful",
      },
      "root": "/Users/kara.brightwell/Work/dotcom-tool-kit/packages/n-test/lib/index.js",
    },
    "../../../../npm": Object {
      "PackageJsonLifecycleInstaller": [Function],
      "id": "../../../../npm",
      "lifecycles": Object {
        "build:local": [Function],
        "test:local": [Function],
      },
      "parent": Object {
        "id": "successful test root",
        "root": "packages/cli/test/files/successful",
      },
      "root": "/Users/kara.brightwell/Work/dotcom-tool-kit/packages/npm/lib/index.js",
    },
    "@dotcom-tool-kit/state": Object {
      "id": "@dotcom-tool-kit/state",
      "parent": Object {
        "id": "../../../../circleci",
        "lifecycles": Object {
          "build:ci": [Function],
          "test:ci": [Function],
          "test:remote": [Function],
        },
        "parent": Object {
          "id": "successful test root",
          "root": "packages/cli/test/files/successful",
        },
        "root": "/Users/kara.brightwell/Work/dotcom-tool-kit/packages/circleci/lib/index.js",
      },
      "readState": [Function],
      "root": "/Users/kara.brightwell/Work/dotcom-tool-kit/packages/state/lib/index.js",
      "writeState": [Function],
    },
  },
  "root": "/Users/kara.brightwell/Work/dotcom-tool-kit/packages/cli",
}
`;
